go/types/struct.go

```diff
@@ -131,180 +182,290 @@ func (s Struct) Hash() hash.Hash {
 // struct.
 func (s Struct) IterFields(cb func(name string, value Value)) {
-	for i := 0; i < len(s.fieldNames); i++ {
-		cb(s.fieldNames[i], s.values[i])
+	dec, count := s.decoderSkipToFields()
+	for i := uint64(0); i < count; i++ {
+		cb(dec.readString(), dec.readValue())
 	}
 }
 
-func (s Struct) Kind() NomsKind {
-	return StructKind
+type structPartCallbacks interface {
+	name(n string)
+	count(c uint64)
+	fieldName(n string)
+	fieldValue(v Value)
+	end()
 }
 
-// MaybeGet returns the value of a field in the struct. If the struct does not a have a field with
-// the name name then this returns (nil, false).
-func (s Struct) MaybeGet(n string) (Value, bool) {
-	i := s.findField(n)
-	if i == -1 {
-		return nil, false
+func (s Struct) iterParts(cbs structPartCallbacks) {
+	dec := s.decoder()
+	dec.skipKind()
+	cbs.name(dec.readString())
+	count := dec.readCount()
+	cbs.count(count)
+	for i := uint64(0); i < count; i++ {
+		cbs.fieldName(dec.readString())
+		cbs.fieldValue(dec.readValue())
 	}
-
-	return s.values[i], true
+	cbs.end()
 }
 
-func (s Struct) searchField(name string) int {
-	return sort.Search(len(s.fieldNames), func(i int) bool { return s.fieldNames[i] >= name })
+func (s Struct) Kind() NomsKind {
+	return StructKind
 }
 
-func (s Struct) findField(name string) int {
-	i := s.searchField(name)
-	if i == len(s.fieldNames) || s.fieldNames[i] != name {
-		return -1
+// MaybeGet returns the value of a field in the struct. If the struct does not a have a field with
+// the name name then this returns (nil, false).
+func (s Struct) MaybeGet(n string) (v Value, found bool) {
+	dec, count := s.decoderSkipToFields()
+	for i := uint64(0); i < count; i++ {
+		name := dec.readString()
+		if name == n {
+			found = true
+			v = dec.readValue()
+			return
+		}
+		if name > n {
+			return
+		}
+		dec.skipValue()
 	}
-	return i
+
+	return
 }
 
 // Get returns the value of a field in the struct. If the struct does not a have a field with the
 // name name then this panics.
 func (s Struct) Get(n string) Value {
-	i := s.findField(n)
-	if i == -1 {
+	v, ok := s.MaybeGet(n)
+	if !ok {
 		d.Chk.Fail(fmt.Sprintf(`Struct has no field "%s"`, n))
 	}
-	return s.values[i]
+	return v
 }
 
 // Set returns a new struct where the field name has been set to value. If name is not an
 // existing field in the struct or the type of value is different from the old value of the
 // struct field a new struct type is created.
 func (s Struct) Set(n string, v Value) Struct {
-	i := s.searchField(n)
-
-	if i != len(s.fieldNames) && s.fieldNames[i] == n {
-		// Found
-		values := make([]Value, len(s.fieldNames))
-		copy(values, s.values)
-		values[i] = v
-
-		// No need to validate
-		return newStruct(s.name, s.fieldNames, values)
+	verifyFieldName(n)
+	return s.set(n, v, 0)
+}
+
+func (s Struct) set(n string, v Value, addedCount int) Struct {
+	// TODO: Reuse bytes if we end up adding a field
+	dec := s.decoder()
```

Comment:
How about having `set` take a nomsWriter? That way if the optimistic path fails, you can just `reset` the writer and reuse it. That way it avoids creating a second buffer allocation.

Reply:
That is a good idea